source/adaptive_load/adaptive_load_controller_impl.cc

```diff
@@ -0,0 +1,444 @@
+    ENVOY_LOG_MISC(info, "Evaluation: {}", evaluation.DebugString());
+  }
+  return result;
+}
+
+/**
+ * Sets the output code and message in the adaptive load session output.
+ *
+ * @param code A Google status code, such as ::grpc::DEADLINE_EXCEEDED.
+ * @param message Status message text
+ * @param output Adaptive load output proto
+ */
+void SetOutputError(int code, absl::string_view message, AdaptiveLoadSessionOutput& output) {
+  output.mutable_session_status()->set_code(code);
+  output.mutable_session_status()->set_message(std::string(message));
+  ENVOY_LOG_MISC(info, message);
+}
+
+/**
+ * Returns a copy of the input spec with default values inserted.
+ *
+ * @param original_spec Valid adaptive load session spec.
+ *
+ * @return Adaptive load session spec with default values inserted.
+ */
+AdaptiveLoadSessionSpec SetDefaults(const AdaptiveLoadSessionSpec& original_spec) {
+  AdaptiveLoadSessionSpec spec = original_spec;
+  if (!spec.has_measuring_period()) {
+    spec.mutable_measuring_period()->set_seconds(10);
+  }
+  if (!spec.has_convergence_deadline()) {
+    spec.mutable_convergence_deadline()->set_seconds(300);
+  }
+  if (!spec.has_testing_stage_duration()) {
+    spec.mutable_testing_stage_duration()->set_seconds(30);
+  }
+  for (nighthawk::adaptive_load::MetricSpecWithThreshold& threshold :
+       *spec.mutable_metric_thresholds()) {
+    if (threshold.metric_spec().metrics_plugin_name().empty()) {
+      threshold.mutable_metric_spec()->set_metrics_plugin_name("nighthawk.builtin");
+    }
+    if (!threshold.threshold_spec().has_weight()) {
+      threshold.mutable_threshold_spec()->mutable_weight()->set_value(1.0);
+    }
+  }
+  for (nighthawk::adaptive_load::MetricSpec& metric_spec :
+       *spec.mutable_informational_metric_specs()) {
+    if (metric_spec.metrics_plugin_name().empty()) {
+      metric_spec.set_metrics_plugin_name("nighthawk.builtin");
+    }
+  }
+  return spec;
+}
+
+/**
+ * Checks whether a session spec is valid: No forbidden fields in Nighthawk traffic spec; no bad
+ * plugin references or bad plugin configurations (step controller, metric, scoring function); no
+ * nonexistent metric names. Reports all errors in one pass.
+ *
+ * @param spec A potentially invalid adaptive load session spec.
+ *
+ * @return Status OK if no problems were found, or InvalidArgument with all errors.
+ */
+absl::Status CheckSessionSpec(const nighthawk::adaptive_load::AdaptiveLoadSessionSpec& spec) {
+  std::vector<std::string> errors;
+  if (spec.nighthawk_traffic_template().has_duration()) {
+    errors.emplace_back(
+        "nighthawk_traffic_template should not have |duration| set. Set |measuring_period| "
+        "and |testing_stage_duration| in the AdaptiveLoadSessionSpec proto instead.");
+  }
+  if (spec.nighthawk_traffic_template().has_open_loop()) {
+    errors.emplace_back(
+        "nighthawk_traffic_template should not have |open_loop| set. Adaptive Load always "
```

Comment:
I may misremember, but open loop needs to be *set* to avoid closed loop mode?
One tought I have here is that there are OSS load gens out there that don't have an open
loop mode, or have one with a partial feedback loop (e.g. when hitting connection limits). That might
be worth considering in the context of enforcing this, if encouraging other implementations of the 
adaptive load controllers is desired?

Reply:
I definitely meant to *enable* open loop mode automatically and not allow mentioning it in the input. The intent here was to avoid confusion by not letting the open loop option even appear in example protos, whether true or false, since we force  open loop = true anyway.

Reply:
Thanks for the context; still had a question about the rationale around this which I just asked in a comment higher up. Resolving this one to deduplicate / consolidate the discussion on a single thread.